cranelift/codegen/src/isa/riscv64/inst/encode.rs

```diff
@@ -334,6 +334,57 @@ pub fn encode_cr_type(op: CrOp, rd: WritableReg, rs2: Reg) -> u16 {
     bits |= unsigned_field_width(op.op().bits(), 2);
     bits |= reg_to_gpr_num(rs2) << 2;
     bits |= reg_to_gpr_num(rd.to_reg()) << 7;
-    bits |= unsigned_field_width(op.funct4(), 5) << 12;
+    bits |= unsigned_field_width(op.funct4(), 4) << 12;
+    bits.try_into().unwrap()
+}
+
+// This isn't technically a instruction format that exists. It's just a CR type
+// where the source is rs1, rs2 is zero. rs1 is never written to.
+//
+// Used for C.JR and C.JALR
+pub fn encode_cr2_type(op: CrOp, rs1: Reg) -> u16 {
+    encode_cr_type(op, WritableReg::from_reg(rs1), zero_reg())
+}
+
+// Encode a CA type instruction.
+//
+// 0--1-2-----4-5--------6-7--------9-10------15
+// |op |  rs2  |  funct2  |  rd/rs1  | funct6 |
+pub fn encode_ca_type(op: CaOp, rd: WritableReg, rs2: Reg) -> u16 {
+    let mut bits = 0;
+    bits |= unsigned_field_width(op.op().bits(), 2);
+    bits |= reg_to_compressed_gpr_num(rs2) << 2;
+    bits |= unsigned_field_width(op.funct2(), 2) << 5;
+    bits |= reg_to_compressed_gpr_num(rd.to_reg()) << 7;
+    bits |= unsigned_field_width(op.funct6(), 6) << 10;
+    bits.try_into().unwrap()
+}
+
+// Encode a CJ type instruction.
+//
+// The imm field is a 11 bit signed immediate that is shifted left by 1.
+//
+// 0--1-2-----12-13--------15
+// |op |  imm   |  funct3  |
+pub fn encode_cj_type(op: CjOp, imm: Imm12) -> u16 {
+    let imm = imm.as_u32();
+    debug_assert!(imm & 1 == 0);
+
+    // The offset bits are in rather weird positions.
+    // [11|4|9:8|10|6|7|3:1|5]
+    let mut imm_field = 0;
+    imm_field |= ((imm >> 11) & 1) << 10;
+    imm_field |= ((imm >> 4) & 1) << 9;
+    imm_field |= ((imm >> 8) & 3) << 7;
+    imm_field |= ((imm >> 10) & 1) << 6;
+    imm_field |= ((imm >> 6) & 1) << 5;
+    imm_field |= ((imm >> 7) & 1) << 4;
+    imm_field |= ((imm >> 1) & 7) << 1;
+    imm_field |= ((imm >> 5) & 1) << 0;
```

Comment:
Oh wow this is something else... I'm curious but do you know if there's a link explaining why the bits are this way? Mostly I'm just curious!

Reply:
There's some discussion of this in [Andrew Waterman's PhD thesis](https://www2.eecs.berkeley.edu/Pubs/TechRpts/2016/EECS-2016-1.html) (his thesis work was the ISA design) -- e.g. on page 31 of that PDF he discusses how certain immediate bits always come from the same position in instruction formats and this reduces MUX logic overhead. Kind of a fascinating design point actually (how few gates can you use for an RV32I implementation).

Reply:
Whoa wild! I've got some reading ahead of me :)